model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto

```diff
@@ -484,24 +522,38 @@ message ProcessBundleSplitResponse {
   // (if the bundle is large) and often a more efficient representation
   // on the runner side (e.g. if the set of elements can be represented
   // as some range in an underlying dataset).
+  //
+  // Note that for a split the following properties must hold:
+  // - last_primary_element < first_residual_element
+  // - primary roots and residual roots can only be specified if the
+  //   last_primary_element + 1 < first_residual_element
+  //   (typically there is one primary and residual root per element in the
+  //   range (last_primary_element, first_residual_element))
+  //
+  // Note that subsequent splits of the same bundle must ensure that:
+  // - the last_primary_element does not increase
```

Comment:
This is not a requirement. The underlying requirement is that primary_n + residual_n = primary_{n-1}.

In practice part or all of last_primary_element + 1 is often part of the previous residual.